sv.c

```diff
@@ -17878,6 +17876,64 @@ Perl_sv_regex_global_pos_clear(pTHX_ SV *sv)
         mg->mg_len = -1;
 }
 
+/*
+=for apidoc sv_vstring_get
+
+If the given SV has vstring magic, stores the string pointer and length into
+the variables addressed by C<pvp> and C<lenp>, and returns true.  If not,
+returns false.
+
+If a pointer is returned to the caller in the C<pvp> variable, it will point
+to memory owned by the SV itself.  The caller is not responsible for freeing
+it after this call, though it will not remain valid for longer than the
+lifetime of the SV itself.  The caller should take a copy of it if it needs
+to be accessed after this time.
+
+=cut
+*/
+
+bool
+Perl_sv_vstring_get(pTHX_ SV * const sv, const char **pvp, STRLEN *lenp)
+{
+    PERL_ARGS_ASSERT_SV_VSTRING_GET;
+
+    MAGIC *mg = SvVSTRING_mg(sv);
+    if(!mg) return false;
+
+    if(pvp)  *pvp  = mg->mg_ptr;
+    if(lenp) *lenp = mg->mg_len;
+    return true;
+}
+
+/*
+=for apidoc sv_vstring_set
+
+Applies vstring magic to the given SV, storing the string given by C<pv> for
+C<len> bytes into it.  If the SV already had vstring magic, the previous
+stored string is freed and replaced.
+
+=cut
```

Comment:
This should probably point at upg_version() and new_version(), which is what you want if you're not using a pre-parsed version.

Reply:
Thinking about this...

Storable uses this mechanism to restore the magic to match an SV it was originally provided, eg. from a `vX.Y`:
```
$ perl -MDevel::Peek -e 'Dump(v1.1.2)'
SV = PVMG(0x556625c6bd00) at 0x556625c29fe0
  REFCNT = 1
  FLAGS = (RMG,POK,IsCOW,READONLY,PROTECT,pPOK)
  IV = 0
  NV = 0
  PV = 0x556625c325e0 "\x01\x01\x02"\0
  CUR = 3
  LEN = 10
  COW_REFCNT = 0
  MAGIC = 0x556625c531d0
    MG_VIRTUAL = 0
    MG_TYPE = PERL_MAGIC_vstring(V)
    MG_LEN = 6
    MG_PTR = 0x556625c1bb60 "v1.1.2"
```
Note that the magic matches the PV.

I do think that the docs should indicate this is only useful when the PV and the magic are kept in sync (or someone is going to see unexpected results).